### PR TITLE
Feature/un 83 Chain reorganization detection

### PR DESCRIPTION
**Pull request by AlexRipoll**

### CODE REVIEW

File: internal/state/sync/ownership/sync.go [hunk 1]
```diff
@@ -51,3 +54,19 @@ func (s *service) GetCurrentOwnershipBlock() (uint64, error) {
...
+	return s.tx.Set([]byte(endRangeBlockHash), blockHash.Bytes())
+}
+
+func (s *service) GetOwnershipEndRangeBlockHash() (common.Hash, error) {
+	value, err := s.tx.Get([]byte(endRangeBlockHash))
+	if err != nil {
+		return common.Hash{}, err
+	}
+	if value == nil {
+		value = common.Hash{}.Bytes()
```
Comment (zorjak) [inline internal/state/sync/ownership/sync.go]: can you return here  `common.Hash{}, nil`?

File: internal/state/sync/evolution/sync.go [hunk 2]
```diff
@@ -50,3 +53,19 @@ func (s *service) GetCurrentEvoBlockTimestamp() (uint64, error) {
...
+	return s.tx.Set([]byte(endRangeBlockHash), blockHash.Bytes())
+}
+
+func (s *service) GetEvoEndRangeBlockHash() (common.Hash, error) {
+	value, err := s.tx.Get([]byte(endRangeBlockHash))
+	if err != nil {
+		return common.Hash{}, err
+	}
+	if value == nil {
+		value = common.Hash{}.Bytes()
```
Comment (zorjak) [inline internal/state/sync/evolution/sync.go]: can you return here `common.Hash{}, nil`?

File: cmd/main.go [hunk 3]
```diff
@@ -239,22 +250,79 @@ func scanUniversalChain(ctx context.Context, c *config.Config, client scan.EthCl
...
-			// load merkle trees for all contracts whose events have to be scanned for
-			// scanning contracts for events on the ownership chain
-			// getting transfer events from scan events
-			// retrieving minted events and update the state accordingly
-			nextStartingBlock, err := scanAndDigest(ctx, stateService, c, s, startingBlock, lastBlock, lastOwnershipBlockTimestamp, client)
+
+			err = processUniversalBlockRange(ctx, c, client, stateService, s, startingBlock, lastBlock, lastOwnershipBlockTimestamp)
 			if err != nil {
+				if errors.Is(err, ReorgError{}) {
+					return err
```
Comment (zorjak) [inline cmd/main.go]: Shouldn't this be Fatal for now? at least until we support reorgs.

File: cmd/main.go [hunk 4]
```diff
@@ -239,22 +250,79 @@ func scanUniversalChain(ctx context.Context, c *config.Config, client scan.EthCl
...
+
+			startingBlock = lastBlock + 1
 		}
 	}
 }
 
+func processUniversalBlockRange(ctx context.Context, c *config.Config, client scan.EthClient, stateService state.Service, s scan.Scanner, startingBlock, lastBlock, lastOwnershipBlockTimestamp uint64) error {
+	tx := stateService.NewTransaction()
+	defer tx.Discard()
+	// retrieve the hash of the final block of the previous iteration.
```
Comment (zorjak) [inline cmd/main.go:270]: maybe this part that detects reorg you can put in a separate function?

**Review by zorjak — APPROVED** [review]

File: cmd/main.go [hunk 5]
```diff
@@ -239,22 +250,79 @@ func scanUniversalChain(ctx context.Context, c *config.Config, client scan.EthCl
...
+	prevLastBlockHash, err := tx.GetOwnershipEndRangeBlockHash()
```
Comment (zorjak) [inline cmd/main.go:280]: maybe this part that detects reorg you can put in a separate function?

**Review by zorjak — CHANGES_REQUESTED** [review]

File: cmd/main.go (hunk 3 above)
Comment (AlexRipoll) [reply cmd/main.go]: No, upon reaching this point, the program will exit the for loop with an error, causing the node to shut down.

File: cmd/main.go [hunk 6]
```diff
@@ -380,53 +438,106 @@ func scanEvoChain(ctx context.Context, c *config.Config, client scan.EthClient,
...
+		var prevIterLastBlock *types.Block
+		prevIterLastBlockNumber := startingBlock - 1
+		prevIterLastBlock, err = client.BlockByNumber(ctx, big.NewInt(int64(prevIterLastBlockNumber)))
+		if err != nil {
+			slog.Error("error occurred while fetching new LaosEvolution start range block", "err", err.Error())
+			return nil
+		}
+
+		if prevIterLastBlock.Hash().Cmp(prevLastBlockHash) != 0 {
+			slog.Error("evolution chain reorganization detected", "block number", startingBlock-1, "chain hash", prevIterLastBlock.Hash().String(), "storage hash", prevLastBlockHash.String())
```
Comment (luispdm) [inline cmd/main.go]: I think we should add further instructions after this message, like: "please, wipe out the database and sync again from the desired block number" (same for the ownership chain reorg detection)

**Review by luispdm — CHANGES_REQUESTED** [review]

**Review by zorjak — APPROVED** [review]

File: cmd/main.go [hunk 7]
```diff
@@ -239,20 +254,90 @@ func scanUniversalChain(ctx context.Context, c *config.Config, client scan.EthCl
...
+func verifyChainConsistency(ctx context.Context, client scan.EthClient, prevLastBlockHash common.Hash, startingBlock uint64) error {
+	// During the initial iteration, no hash is stored in the database, so this code block is bypassed.
+	// Verify whether the hash of the last block from the previous iteration remains unchanged; if it differs,
+	// it indicates a reorganization has taken place.
+	if prevLastBlockHash != (common.Hash{}) {
+		var prevIterLastBlock *types.Block
+		prevIterLastBlockNumber := startingBlock - 1
+		prevIterLastBlock, err := client.BlockByNumber(ctx, big.NewInt(int64(prevIterLastBlockNumber)))
+		if err != nil {
+			slog.Error("error occurred while retrieving new ownership start range block", "err", err.Error())
```
Comment (luispdm) [inline cmd/main.go]: Remove "ownership"

**Review by luispdm — CHANGES_REQUESTED** [review]

**Review by luispdm — APPROVED** [review]

**Pull request merged**